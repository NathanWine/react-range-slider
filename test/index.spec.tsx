import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import RangeSlider from '../src';

const mockOnChange = jest.fn();
const mockOnDragEnd = jest.fn();

function setup(axis: 'x' | 'y' | 'xy' = 'x', props = {}): ReactWrapper {
  return mount(
    <RangeSlider
      axis={axis}
      classNamePrefix="rrs"
      onChange={mockOnChange}
      onDragEnd={mockOnDragEnd}
      {...props}
    />,
  );
}

describe('RangeSlider', () => {
  describe('with axis `x`', () => {
    const wrapper = setup('x' as const);

    it('should render properly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should handle mousedown on the handle', () => {
      // @ts-ignore
      Element.prototype.getBoundingClientRect = () => ({
        bottom: 0,
        height: 20,
        left: 50,
        right: 0,
        top: 50,
        width: 200,
      });

      wrapper.find('.rrs__handle').simulate('mousedown', {
        clientX: 100,
        clientY: 0,
        currentTarget: {},
      });

      // @ts-ignore
      expect(wrapper.instance().offset).toEqual({ x: 100, y: 0 });
    });

    it('should handle mousemove', () => {
      // @ts-ignore
      Element.prototype.getBoundingClientRect = () => ({
        bottom: 0,
        height: 20,
        left: 50,
        right: 0,
        top: 50,
        width: 200,
      });

      document.dispatchEvent(
        new MouseEvent('mousemove', {
          clientX: 230,
          clientY: 0,
        }),
      );

      expect(mockOnChange).toHaveBeenLastCalledWith(
        { x: 65, y: 0 },
        {
          axis: 'x',
          classNamePrefix: 'rrs',
          onChange: expect.any(Function),
          onDragEnd: expect.any(Function),
          xMax: 100,
          xMin: 0,
          xStep: 1,
          yMax: 100,
          yMin: 0,
          yStep: 1,
        },
      );
    });

    it('should handle mouseup', () => {
      document.dispatchEvent(
        new MouseEvent('mouseup', {
          clientX: 230,
          clientY: 0,
        }),
      );

      expect(mockOnDragEnd).toHaveBeenLastCalledWith(
        { x: 65, y: 0 },
        {
          axis: 'x',
          classNamePrefix: 'rrs',
          onChange: expect.any(Function),
          onDragEnd: expect.any(Function),
          xMax: 100,
          xMin: 0,
          xStep: 1,
          yMax: 100,
          yMin: 0,
          yStep: 1,
        },
      );
    });

    it('should handle touchstart on the handle', () => {
      // @ts-ignore
      Element.prototype.getBoundingClientRect = () => ({
        bottom: 0,
        height: 20,
        left: 50,
        right: 0,
        top: 50,
        width: 200,
      });

      wrapper.find('.rrs__handle').simulate('touchstart', {
        clientX: 50,
        clientY: 0,
        currentTarget: {},
      });

      // @ts-ignore
      expect(wrapper.instance().offset).toEqual({ x: 50, y: 0 });
    });

    it('should handle clicks on the track', () => {
      // @ts-ignore
      Element.prototype.getBoundingClientRect = () => ({
        bottom: 0,
        height: 20,
        left: 50,
        right: 0,
        top: 50,
        width: 200,
      });

      wrapper.find('.rrs__track').simulate('click', {
        clientX: 100,
        clientY: 0,
        currentTarget: {},
      });

      expect(mockOnChange).toHaveBeenLastCalledWith(
        { x: 25, y: 0 },
        {
          axis: 'x',
          classNamePrefix: 'rrs',
          onChange: expect.any(Function),
          onDragEnd: expect.any(Function),
          xMax: 100,
          xMin: 0,
          xStep: 1,
          yMax: 100,
          yMin: 0,
          yStep: 1,
        },
      );
    });

    it('should handle focus, keydown and blur', () => {
      wrapper.find('.rrs__handle').simulate('focus');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowRight' }));
      expect(wrapper.state('x')).toBe(26);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowUp' }));
      expect(wrapper.state('x')).toBe(27);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowLeft' }));
      expect(wrapper.state('x')).toBe(26);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state('x')).toBe(25);

      wrapper.find('.rrs__handle').simulate('blur');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state('x')).toBe(25);
    });
  });

  describe('with axis `xy`', () => {
    const wrapper = setup('xy', { x: 20, xMax: 20, y: 20, yMax: 10 });

    it('should render properly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should handle focus, keydown and blur', () => {
      wrapper.find('.rrs__handle').simulate('focus');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowRight' }));
      expect(wrapper.state()).toEqual({ x: 20, y: 10 });

      wrapper.setProps({ y: 10 });

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowUp' }));
      expect(wrapper.state()).toEqual({ x: 20, y: 10 });

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state()).toEqual({ x: 20, y: 9 });

      wrapper.setProps({ y: 9 });

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowLeft' }));
      expect(wrapper.state()).toEqual({ x: 19, y: 9 });

      wrapper.find('.rrs__handle').simulate('blur');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state()).toEqual({ x: 19, y: 9 });
    });
  });

  describe('with axis `y`', () => {
    const wrapper = setup('y');

    it('should render properly', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should handle focus, keydown and blur', () => {
      wrapper.find('.rrs__handle').simulate('focus');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowRight' }));
      expect(wrapper.state('y')).toBe(1);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowUp' }));
      expect(wrapper.state('y')).toBe(2);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowLeft' }));
      expect(wrapper.state('y')).toBe(1);

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state('y')).toBe(0);

      wrapper.find('.rrs__handle').simulate('blur');

      document.dispatchEvent(new KeyboardEvent('keydown', { code: 'ArrowDown' }));
      expect(wrapper.state('y')).toBe(0);
    });
  });
});
