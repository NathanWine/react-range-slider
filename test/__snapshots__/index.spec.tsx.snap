// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider with axis \`x\` should render properly 1`] = `
<div style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 32px; width: 100%;"
     class="rrs"
>
  <div class="rrs__track"
       style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 20px; position: relative; width: 100%;"
       role="presentation"
  >
    <div class="rrs__range"
         style="width: 0%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; height: 100%; top: 0px;"
    >
    </div>
    <div style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 0%; bottom: 0%;"
         role="presentation"
    >
      <span class="rrs__handle"
            style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; height: 26px; left: -5px; top: -3px; width: 10px;"
            tabindex="0"
            role="slider"
            aria-label="slider handle"
            aria-orientation="horizontal"
            aria-valuemin="0"
            aria-valuenow="0"
            aria-valuemax="100"
      >
      </span>
    </div>
  </div>
</div>
`;

exports[`RangeSlider with axis \`xy\` should render properly 1`] = `
<div style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 100%; width: 100%;"
     class="rrs"
>
  <div class="rrs__track"
       style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 100%; position: relative; width: 100%; min-height: 50px;"
       role="presentation"
  >
    <div class="rrs__range"
         style="height: 100%; width: 100%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; bottom: 0px;"
    >
    </div>
    <div style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 100%; bottom: 100%;"
         role="presentation"
    >
      <span class="rrs__handle"
            style="background-color: transparent; border: 2px solid #000; border-radius: 50%; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; bottom: -10px; height: 20px; left: -10px; width: 20px;"
            tabindex="0"
            role="slider"
            aria-label="slider handle"
            aria-valuemin="0"
            aria-valuenow="20"
            aria-valuemax="20"
      >
      </span>
    </div>
  </div>
</div>
`;

exports[`RangeSlider with axis \`y\` should render properly 1`] = `
<div style="box-sizing: border-box; display: inline-block; padding: 6px; transition: height 0.4s, width 0.4s; height: 100%; width: 32px;"
     class="rrs"
>
  <div class="rrs__track"
       style="background-color: rgb(204, 204, 204); border-radius: 3px; box-sizing: border-box; height: 100%; position: relative; width: 20px; min-height: 50px;"
       role="presentation"
  >
    <div class="rrs__range"
         style="height: 0%; background-color: rgb(0, 123, 255); border-radius: 3px; position: absolute; bottom: 0px; left: 0px; width: 100%;"
    >
    </div>
    <div style="box-sizing: border-box; height: 20px; position: absolute; transition: height 0.4s, width 0.4s; width: 20px; left: 0%; bottom: 0%;"
         role="presentation"
    >
      <span class="rrs__handle"
            style="background-color: rgb(255, 255, 255); border: 2px solid #000; border-radius: 4px; box-sizing: border-box; display: block; position: absolute; transition: height 0.4s, width 0.4s; bottom: -5px; height: 10px; left: -3px; width: 26px;"
            tabindex="0"
            role="slider"
            aria-label="slider handle"
            aria-orientation="vertical"
            aria-valuemin="0"
            aria-valuenow="0"
            aria-valuemax="100"
      >
      </span>
    </div>
  </div>
</div>
`;
